src/tir/transforms/storage_rewrite.cc

```diff
@@ -885,108 +893,528 @@ class StoragePlanRewriter : public StmtExprMutator {
+    //     << "Attempting to retrieve " << value_dtype.lanes() << " lanes of data with "
+    //     << index.dtype().lanes() << " indices into an array whose elements have "
+    //     << var_info.element_dtype.lanes() << " lanes.  "
+    //     << "Expected output with " << index.dtype().lanes() * var_info.element_dtype.lanes()
+    //     << " lanes.";
+
+    // If the index is a RampNode with stride of 1 and offset
+    // divisible by the number of number of lanes, and the predicate
+    // does not apply any masking, then this array access could be
+    // vectorized.
+    const RampNode* ramp_index = index.as<RampNode>();
+    if (ramp_index && is_one(ramp_index->stride) && is_one(predicate)) {
+      arith::ModularSet me = analyzer_.modular_set(ramp_index->base);
+      if ((me->coeff % ramp_index->lanes == 0) && (me->base % ramp_index->lanes == 0)) {
+        lanes_used = ramp_index->lanes;
+      }
+    }
+
+    var_info.lanes_used.insert(lanes_used);
   }
 
-  // Internal access map
-  std::unordered_map<const VarNode*, std::vector<DataType> > acc_map_;
-  // Variables to remap
-  Map<tir::Var, PrimExpr> var_remap_;
+  // Map of buffer variable information determined
+  std::unordered_map<const VarNode*, BufferVarInfo> info_map_;
+
+  //
+  bool allow_untyped_pointers_{false};
+
   // internal analyzer
   arith::Analyzer analyzer_;
 };
 
-PrimFunc PointerValueTypeRewrite(PrimFunc f) {
-  auto* n = f.CopyOnWrite();
-  VectorAllocRewriter rewriter;
-  n->body = rewriter(std::move(n->body));
+/* \brief Rewrites buffer/pointer variables from scalar types to vectorized
+ * types.
+ *
+ * Some runtimes do not allow casting between composite types and the underlying
+ * base type (e.g. Vulkan, casting from 1-lane float16* to 4-lane float16x4*).
+ * In these cases, in order to have vectorized load/store on an array, the
+ * element type of that array must be vectorized.  This is in contrast to C-style
+ * runtimes, in which `float16x4* vec = *(float16x4*)(float_arr + offset)` is
+ * valid.
+ *
+ * By default, VectorTypeRewriter will attempt to rewrite all buffer variables to
+ * vectorized access, if the load/store occurring in the PrimFunc are all
+ * vectorized.  This includes adjusting the indices being used to access the
+ * array.  (e.g. If `float16* scalar_arr` is being converted to `float16x4*
+ * vec_arr`, then `scalar_arr[Ramp(offset, 1, 4)]` will be converted to
+ * `vec_arr[offset/4]`.)
+ *
+ * Currently, several of the C-style runtimes do not support buffers whose
+ * elements are vectorized types, or rely on the presence of the Ramp nodes to
+ * identify vectorized loads.  The boolean parameters in the constructor are to
+ * mimic the previous behavior of VectorTypeRewriter, to avoid breaking these
+ * runtimes.  Once all runtimes support vectorized buffer elements, these
+ * parameters can be removed.
+ */
+class VectorTypeRewriter : public StmtExprMutator {
```

Comment:
I think I understand what this pass does, but I don't get the following point:

My understanding is that
* For SPIR-V target, this pass should be run
* For C-codegen targets, this should be nop

How these two conflicting requirements are satisfied?

Reply:
Ah, good point.  Currently, the `tir.transform.PointerValueTypeRewrite` pass is only used in the SPIR-V codegen step, while the `tir.transform.StorageRewrite` pass is used as part of [`tvm.lower`](https://github.com/apache/tvm/blob/main/src/driver/driver_api.cc#L242).  These two passes both share the `PointerValueTypeRewrite` function.  Currently, those two conflicting requirements are handled by having different arguments to `PointerValueTypeRewrite` for those two cases.

Unfortunately, for the C-codegen, it isn't quite a nop, as that would have been simpler to handle.  Instead, the `AllocateNode`, any function parameters, and references to those variables get rewritten, but the `StoreNode` and `LoadNode` do not have their indices rewritten to account for the different variable type.  I'm [trying to determine](https://discuss.tvm.apache.org/t/pre-rfc-vectorized-tir-buffers/10615/4) the best way to handle that on the C codegens, but for now the previous behavior can be maintained with the boolean options to `PointerValueTypeRewrite`.

Reply:
Oh I didn't know that `PointerValueTypeRewrite` is called from `build_vulkan.cc` !

Reply:
I was surprised as well, as my mental model was that all changes to the TIR graph occur before being passed to the codegen.  This extra call is fairly vulkan-specific, and from talking with @tqchen exists to adjust the pointer-type through which Vulkan accesses an array.  For cuda, it's unnecessary because the pointer can be cast to the desired output type, but Vulkan doesn't allow those pointer casts.  Instead, we need to choose one specific type for each pointer that is passed in, and stick with it through the entire `PrimFunc`.

My preference would be to pull it out into a target-aware optimization pass, which could be added into either `tvm.lower` or `tvm.build`.  I've put together some comments over [on discuss](https://discuss.tvm.apache.org/t/pre-rfc-vectorized-tir-buffers/10615/4), and will tag you there as well.